### PR TITLE
Login screen

### PR DESCRIPTION
**Pull request by anastasiaShpakava**

### CODE REVIEW

**Review by pcfaktor — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -69,6 +71,8 @@ moshi-codegen = { module = "com.squareup.moshi:moshi-kotlin-codegen", version.re
 androidx-room-runtime = { module = "androidx.room:room-runtime", version.ref = "room" }
 androidx-room-compiler = { module = "androidx.room:room-compiler", version.ref = "room" }
 androidx-room-ktx = { module = "androidx.room:room-ktx", version.ref = "room" }
+androidx-compose-desugar = { module = "com.android.tools:desugar_jdk_libs", version.ref = "composeDesugar" }
```
Comment (bstolyarov) [inline gradle/libs.versions.toml:74]: Why do we need it?

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -4,4 +4,27 @@
...
+    <string name="required">Required</string>
+    <string name="empty_fields_error">Please, fill all fields</string>
+    <string name="problem_error">You have a problem</string>
+    <string name="sign_up">Sign up</string>
+    <string name="country">Country</string>
+    <string name="email_error">Not match</string>
+    <string name="email_not_valid">Email is not valid</string>
+    <string name="registration_password_condition1"> at least 6 characters</string>
+    <string name="registration_password_condition2">at least 1 uppercase letter</string>
+    <string name="birth_day">Birth day</string>
```
Comment (bstolyarov) [inline app/src/main/res/values/strings.xml:21]: It's probably unused

**Review by bstolyarov — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/softteco/template/data/profile/dto/CreateUserDto.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.softteco.template.data.profile.dto
+
+import com.squareup.moshi.Json
+
+data class CreateUserDto(
+    @field:Json(name = "firstName") val firstName: String,
```
[the remark this answers is not in the document]
Comment (pcfaktor) [reply app/src/main/kotlin/com/softteco/template/data/profile/dto/CreateUserDto.kt:6]: I'm not sure if this has had time to become standard.) Anyway, it is out of the scope of PR

File: app/src/main/kotlin/com/softteco/template/data/profile/dto/LoginAuthDto.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.softteco.template.data.profile.dto
+
+import com.squareup.moshi.Json
+
+data class LoginAuthDto(
```
[the remark this answers is not in the document]
Comment (pcfaktor) [reply app/src/main/kotlin/com/softteco/template/data/profile/dto/LoginAuthDto.kt:5]: It is good practice to label fields with their original names. We have no control over the names on the backend side. Sometimes they don't conform to our naming convention or they don't fit for some other reason. There is also the possibility that someone could rename a field (DTO), which could cause problems.

File: app/src/main/kotlin/com/softteco/template/ui/feature/login/LoginScreen.kt [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+@Composable
+private fun ScreenContent(
+	state: LoginViewModel.State,
+	modifier: Modifier = Modifier,
+	onBackClicked: () -> Unit = {},
+	onLoginClicked: () -> Unit = {},
+	onSignUpClicked: () -> Unit = {}
+) {
+	TextSnackbarContainer(
```
[the remark this answers is not in the document]
Comment (pcfaktor) [reply app/src/main/kotlin/com/softteco/template/ui/feature/login/LoginScreen.kt:71]: The component was intended that way. It encapsulates inside itself the box with all screen contents and a SnackbarHost above it

File: app/src/main/kotlin/com/softteco/template/ui/feature/ValidateFields.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import androidx.annotation.StringRes
+import com.softteco.template.Constants
+
+object ValidateFields {
+
+	fun String.isEmailCorrect(): Boolean {
+		return this.matches(Regex(Constants.EMAIL_PATTERN))
+	}
+
+	fun String.isFieldEmpty(): Boolean {
```
[the remark this answers is not in the document]
Comment (pcfaktor) [reply app/src/main/kotlin/com/softteco/template/ui/feature/ValidateFields.kt:12]: Discussed in the background

File: app/src/main/kotlin/com/softteco/template/ui/feature/signUp/SignUpViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	): Flow<State> {
+		return combine(
+			loadingFlow,
+			registrationStateFlow,
+			userNameFlow,
+			emailFlow,
+			passwordFlow,
+			snackBarFlow
+		) { array ->
+			val loading = array[0] as Boolean
```
[the remark this answers is not in the document]
Comment (pcfaktor) [reply app/src/main/kotlin/com/softteco/template/ui/feature/signUp/SignUpViewModel.kt:56]: Discussed in the background. Utils ```combine``` fun for more parameters will be implemented.

File: app/src/main/kotlin/com/softteco/template/ui/theme/Dimens.kt [hunk 8]
```diff
@@ -4,10 +4,9 @@ import androidx.compose.ui.unit.Dp
 import androidx.compose.ui.unit.dp
 
 object Dimens {
-
     val PaddingSmall: Dp = 8.dp
-
     val PaddingNormal: Dp = 16.dp
-
     val PaddingLarge: Dp = 24.dp
+    val PaddingMiddle: Dp = 40.dp
```
[the remark this answers is not in the document]
Comment (pcfaktor) [reply app/src/main/kotlin/com/softteco/template/ui/theme/Dimens.kt:10]: Discussed in the background

File: app/src/main/kotlin/com/softteco/template/data/profile/dto/CreateUserDto.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (bstolyarov) [reply app/src/main/kotlin/com/softteco/template/data/profile/dto/CreateUserDto.kt:6]: It's not a standard for development, I mean it's from JetBrains so it's something natural, default I would say. 
Agree that it's out of scope, let's discuss it separately.

File: app/src/main/kotlin/com/softteco/template/data/profile/dto/LoginAuthDto.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (bstolyarov) [reply app/src/main/kotlin/com/softteco/template/data/profile/dto/LoginAuthDto.kt:5]: If the backend has a different field name we can add annotation for this single field. 
The same as if a backend changes a field name, we also can add an annotation for the updated field, without changing dto field. 
In cases when 99% of fields are equal, annotations look boilerplate. 
Can discuss it in the background :)

**Review by pcfaktor — APPROVED** [review]
Good job!

File: app/src/main/kotlin/com/softteco/template/ui/feature/login/LoginViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+			},
+			onLoginClicked = ::onLogin,
+		)
+	}.stateIn(
+		viewModelScope,
+		SharingStarted.Lazily,
+		State()
+	)
+
+	private fun handleError(error: ErrorEntity) {
```
[the remark this answers is not in the document]
Comment (bstolyarov) [reply app/src/main/kotlin/com/softteco/template/ui/feature/login/LoginViewModel.kt]: Let's do it in separate PR

**Review by bstolyarov — APPROVED** [review]
@anastasiaShpakava Great job! 
@pcfaktor Thank you!

**Pull request merged**